File: micro/loop_overhead.c
// C equivalent of loop_overhead.hl.
// XOR accumulation defeats DCE without overflow concerns; getpid()
// seeds iters/acc so clang can't constant-fold the loop. Caveat:
// clang may vectorize the XOR loop (fair — Hale lowers via LLVM
// too; Go does not auto-vectorize).
#include <stdio.h>
#include <time.h>
#include <unistd.h>

static long long now_ns(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (long long)ts.tv_sec * 1000000000LL + (long long)ts.tv_nsec;
}

int main(void) {
    long long iters = 100000000 + getpid();
    long long t0 = now_ns();
    long long acc = getpid();
    for (long long i = 0; i < iters; i++) {
        acc = acc ^ i;
    }
    long long elapsed = now_ns() - t0;
    printf("iters=%lld\n", iters);
    printf("acc=%lld\n", acc);
    printf("elapsed_ns=%lld\n", elapsed);
    return 0;
}

File: micro/loop_overhead.rs
// Rust equivalent of loop_overhead.hl.
// XOR accumulation defeats DCE without overflow concerns; pid-seeded
// bounds defeat constant folding. LLVM may vectorize the XOR loop
// where Go doesn't — this measures best-effort loop throughput per
// language, same as the .go.

use std::time::Instant;

fn main() {
    let pid = std::process::id() as i64;
    let iters: i64 = 100000000 + pid;
    let t0 = Instant::now();
    let mut acc: i64 = pid;
    for i in 0..iters {
        acc ^= i;
    }
    let elapsed = t0.elapsed().as_nanos();
    println!("iters={}", iters);
    println!("acc={}", acc);
    println!("elapsed_ns={}", elapsed);
}
